payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java

```diff
@@ -491,45 +491,66 @@ public boolean apply(final PaymentTransactionModelDao curPaymentTransactionModel
+                    pastPaymentAttempt.getPaymentMethodId(),
+                    pastPaymentAttempt.getId(),
+                    pastPaymentAttempt.getCreatedDate(),
+                    pastPaymentAttempt.getUpdatedDate(),
+                    pastPaymentAttempt.getCreatedDate(),
+                    pastPaymentAttempt.getPaymentExternalKey(),
+                    pastPaymentAttempt.getTransactionId(),
+                    pastPaymentAttempt.getTransactionExternalKey(),
+                    pastPaymentAttempt.getTransactionType(),
+                    pastPaymentAttempt.getStateName(),
+                    pastPaymentAttempt.getAmount(),
+                    pastPaymentAttempt.getCurrency(),
+                    pastPaymentAttempt.getPluginName(),
+                    (lastPaymentTransaction.getPaymentInfoPlugin() != null) ? lastPaymentTransaction.getPaymentInfoPlugin().getProperties() : null);
+            paymentAttempts.add(paymentAttempt);
         }
 
         // Get Future Payment Attempts from Notification Queue and add them to the list
         try {
-            final NotificationQueue retryQueue = notificationQueueService.getNotificationQueue(DefaultPaymentService.SERVICE_NAME, "retry");
+            final NotificationQueue retryQueue = notificationQueueService.getNotificationQueue(DefaultPaymentService.SERVICE_NAME, DefaultRetryService.QUEUE_NAME);
             final List<NotificationEventWithMetadata<NotificationEvent>> notificationEventWithMetadatas =
                     retryQueue.getFutureNotificationForSearchKeys(internalTenantContext.getAccountRecordId(), internalTenantContext.getTenantRecordId());
 
             for (NotificationEventWithMetadata<NotificationEvent> notificationEvent : notificationEventWithMetadatas) {
                 DefaultPaymentAttempt futurePaymentAttempt = new DefaultPaymentAttempt(
-                        null, //accountId,
-                        null, //paymentMethodId,
-                        ((PaymentRetryNotificationKey) notificationEvent.getEvent()).getAttemptId(), //id,
-                        notificationEvent.getEffectiveDate(), //createdDate,
-                        notificationEvent.getEffectiveDate(), //updatedDate,
-                        null, //paymentExternalKey,
-                        null, //transactionId,
-                        null, //transactionExternalKey,
-                        null, //transactionType,
-                        "SCHEDULED", //stateName,
-                        null, //amount,
-                        null, //currency,
-                        ((PaymentRetryNotificationKey) notificationEvent.getEvent()).getPaymentControlPluginNames().get(0), //pluginName,
-                        null);//pluginProperties
-                result.add(futurePaymentAttempt);
+                        lastPaymentAttemptModelDao.getAccountId(), // accountId
+                        lastPaymentAttemptModelDao.getPaymentMethodId(), // paymentMethodId
```

Comment:
Yep something like that. I would not look at `createdDate` or `updatedDate`, but i would iterate over `pastPaymentAttempts ` starting from last one (DB query already orders them), and stopping when i find the first match (based on `paymentAttemptId`). Actually i am not even sure this is necessary to go reverse order because there should be only one match, but just for safety maybe.